### PR TITLE
Allow specifying explicit product destinations in PackageSettings

### PR DESCRIPTION
**Pull request by fortmarek**
Resolves https://github.com/tuist/tuist-cloud/issues/153

### Short description 📝

Let's take the following `Package.swift`:
```swift
import PackageDescription

let package = Package(
    name: "MyPackage",
    products: [
        .library(
            name: "MyPackage",
            targets: ["MyPackage"]
        ),
        .library(
            name: "MyUIKitPackage",
            targets: ["MyUIKitPackage"]
        ),
    ],
    targets: [
        .target(
            name: "MyPackage"
        ),
        .target(
            name: "MyUIKitPackage"
        ),
    ]
)
```

In this example, we have two products – `MyPackage` and `MyUIKitPackage`. `MyPackage` supports _all_ destinations, whereas `MyUIKitPackage` only support iOS-based platforms since it depends on `UIKit` (which is not available on platforms like macOS).

If you try to use e.g. `tuist cache` when developing a package, we will try to cache _all_ platforms since packages implicitly support all platforms. To explicitly encode in `Package.swift` if some of your products support only _some_ destinations, we're adding a new property of `PackageSettings` called `productDestinations`. For the above example, you could encode that `MyUIKitPackage` supports only iOS-based platforms by:
```swift
#if TUIST
    import ProjectDescription

    let packageSettings = PackageSettings(
        productDestinations: [
            "MyUIKitPackage": [
                .iPad,
                .iPhone,
            ],
        ]
    )
#endif
```

### How to test the changes locally 🧐

Run `tuist generate` in the `spm_package` fixture. `MyPackage` target should have all destinations whereas `MyUIKitPackage` only `iPhone` and `iPad`.

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint:fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [x] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Comment by danieleformichelli** [comment]
Aren't target platform determined depending on which targets depend on the product?

**Comment by pepicrft** [comment]
> Aren't target platform determined depending on which targets depend on the product?

This is where you have a Tuist project depending on Swift targets. I believe the scenario @fortmarek addresses here is where you run Tuist against a Swift Package. In that scenario, we have no information that allows us to narrow-down the list of platforms.

File: Tests/TuistAcceptanceTests/EditAcceptanceTests.swift [hunk 1]
```diff
@@ -40,6 +40,14 @@ final class EditAcceptanceTestAppWithSPMDependencies: TuistAcceptanceTestCase {
     }
 }
 
+final class EditAcceptanceTestSPMPackage: TuistAcceptanceTestCase {
```
Comment (pepicrft) [inline Tests/TuistAcceptanceTests/EditAcceptanceTests.swift:43]: Good one!

File: Tests/TuistKitTests/Services/InstallServiceTests.swift [hunk 2]
```diff
@@ -53,6 +53,8 @@ final class InstallServiceTests: TuistUnitTestCase {
         // Given
         let stubbedPath = try temporaryPath()
 
+        manifestFilesLocator.locatePackageManifestStub = stubbedPath.appending(components: "Tuist", "Package.swift")
```
Comment (pepicrft) [inline Tests/TuistKitTests/Services/InstallServiceTests.swift:56]: Have you considered taking the opportunity to use `@Mockable` with `ManifestsFileLocating`?

**Review by pepicrft — APPROVED** [review]

File: Sources/TuistKit/ProjectEditor/ProjectEditor.swift [hunk 3]
```diff
@@ -175,7 +175,7 @@ final class ProjectEditor: ProjectEditing {
 
         /// We error if the user tries to edit a project in a directory where there are no editable files.
         if projectManifests.isEmpty, editablePluginManifests.isEmpty, helpers.isEmpty, templateSources.isEmpty,
-           resourceSynthesizers.isEmpty, stencils.isEmpty
+           resourceSynthesizers.isEmpty, stencils.isEmpty, packageManifestPath == nil
```
[the remark this answers is not in the document]
Comment (fortmarek) [reply Sources/TuistKit/ProjectEditor/ProjectEditor.swift:178]: So users can run `tuist edit` on a project that has the `Package.swift` manifest only

**Comment by fortmarek** [comment]
> I believe the scenario @fortmarek addresses here is where you run Tuist against a Swift Package.

Yup, that's exactly the scenario this PR aims to fix 🙏

**Comment by danieleformichelli** [comment]
Also, I need we need to support `tuist edit`ing Package.swift file alone if we want to go down this path

**Comment by fortmarek** [comment]
> I see, are we ignoring the setting in case we are running against a tuist project instead?

No, we're not. I feel it would be weird if that setting didn't do anything depending on if the `Package.swift` is a root one or not.

> Also, I need we need to support tuist editing Package.swift file alone if we want to go down this path, right?
(Maybe this PR does it already and I'm missing it)

Yes, the PR already adds support for that.

**Comment by danieleformichelli** [comment]
> No, we're not. I feel it would be weird if that setting didn't do anything depending on if the `Package.swift` is a root one or not.
> 

I think we should.
The fact that we need this for Package.swift generation shouldn't impact (and I would say worsen) the "standard" generation

**Review by danieleformichelli — APPROVED** [review]

**Comment by fortmarek** [comment]
>  The fact that we need this for Package.swift generation shouldn't impact (and I would say worsen) the "standard" generation

Sure, done!

**Pull request merged**